src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3888,61 +3905,69 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
             EnterRegion(finallyRegion);
             AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block));
 
-            if (!knownToImplementIDisposable)
-            {
-                Debug.Assert(!isNotNullableValueType(resource.Type));
-                resource = ConvertToIDisposable(resource, iDisposable, isTryCast: true);
-                int captureId = GetNextCaptureId(finallyRegion);
-                AddStatement(new FlowCaptureOperation(captureId, resource.Syntax, resource));
-                resource = GetCaptureReference(captureId, resource);
-            }
-
-            if (!knownToImplementIDisposable || !isNotNullableValueType(resource.Type))
-            {
-                IOperation condition = MakeIsNullOperation(OperationCloner.CloneOperation(resource));
-                ConditionalBranch(condition, jumpIfTrue: true, endOfFinally);
-                _currentBasicBlock = null;
-            }
-
-            if (!iDisposable.Equals(resource.Type))
+            if (disposeMethod is object)
             {
-                resource = ConvertToIDisposable(resource, iDisposable);
-            }
-
-            AddStatement(tryDispose(resource) ??
-                         MakeInvalidOperation(type: null, resource));
+                ITypeSymbol iDisposable = isAsynchronous
+                                          ? _compilation.CommonGetWellKnownType(WellKnownType.System_IAsyncDisposable).GetITypeSymbol()
+                                          : _compilation.GetSpecialType(SpecialType.System_IDisposable);
 
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.ConsiderEverything);
+                var isNull = resource.GetConstantValue() == ConstantValue.Null;
+                var requiresRuntimeConversion = !isPatternDispose
+                                                && !isAsynchronous
+                                                && !isNull
```

Comment:
>!isNull [](start = 51, length = 7)

It is not clear to me where this null check came from.

Reply:
The VB version of `Using Nothing ... End Using` ends up putting the type of the operation as `Object`, meaning the logic here was attempting to perform a runtime check for it. 

As we're restoring the `knownToImplementIDisposable` members, it's not needed anyway. Agree we should be trying to calculate the minimal amount of things for this exact reason.

Reply:
> The VB version of Using Nothing ... End Using ends up putting the type of the operation as Object, meaning the logic here was attempting to perform a runtime check for it. 

Is addition of this check going to cause a behavior change for C# scenarios? This is exactly the kind of issue that I don't want us to deal here. 

---
In reply to: [528045939](https://github.com/dotnet/roslyn/pull/49481#discussion_r528045939) [](ancestors = 528045939)